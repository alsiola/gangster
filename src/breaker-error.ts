export class BreakerError extends Error {
    constructor(msg: string, public internalError?: any) {
        super(msg);
    }
}
