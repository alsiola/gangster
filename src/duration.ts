export enum DurationInterval {
    Milliseconds = "Milliseconds",
    Seconds = "Seconds",
    Minutes = "Minutes"
}

export interface Duration {
    interval: DurationInterval;
    value: number;
}

export const milliseconds = (value: number): Duration => ({
    interval: DurationInterval.Milliseconds,
    value
});

export const seconds = (value: number): Duration => ({
    interval: DurationInterval.Seconds,
    value
});

export const minutes = (value: number): Duration => ({
    interval: DurationInterval.Minutes,
    value
});

export const toMilliseconds = (duration: Duration): number => {
    switch (duration.interval) {
        case DurationInterval.Milliseconds:
            return duration.value;
        case DurationInterval.Minutes:
            return 60 * duration.value * MS_PER_SEC;
        case DurationInterval.Seconds:
            return duration.value * MS_PER_SEC;
    }
};

const MS_PER_SEC = 1000;
