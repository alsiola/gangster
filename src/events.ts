export enum BreakerEventNames {
    called = "called",
    validResult = "validResult",
    invalidResult = "invalidResult",
    validError = "validError",
    invalidError = "invalidError"
}

export interface BreakerEvents<T, U> {
    [BreakerEventNames.called]: {
        args: T;
    };
}
