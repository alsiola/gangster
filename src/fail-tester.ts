export interface FailureConditions<U> {
    testError?: (a: any) => boolean;
    testResult?: (a: U) => boolean;
}

/**
 * Provides checking capability for failure through two
 * user-provided failureConditions which are simple predicates
 * on U
 */
export class FailTester<U> {
    constructor(private failureConditions: FailureConditions<U> = {}) {}

    /**
     * Is a result a failure case?
     */
    public isFailureResult = (a: U): boolean => {
        return this.failureConditions.testResult
            ? this.failureConditions.testResult(a)
            : false;
    };

    /**
     * Is an error a failure case?
     * If no specific test, then it is.
     */
    public isFailureError = (a: any): boolean => {
        return this.failureConditions.testError
            ? this.failureConditions.testError(a)
            : true;
    };
}
