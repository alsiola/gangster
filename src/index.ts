export { BreakerError } from "./breaker-error";
export { BreakerOpts } from "./breaker";
export { Circuits } from "./circuits";
export { seconds, minutes, Duration, DurationInterval } from "./duration";
export { FailTester, FailureConditions } from "./fail-tester";
export { TripperOpts } from "./tripper";
export * from "./types";
