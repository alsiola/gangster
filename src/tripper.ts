import { Duration, toMilliseconds, seconds } from "./duration";

export interface CreateDefaultTripperOpts {
    threshold: number;
    within: Duration;
    for: Duration;
    outcomeCleanupInterval?: Duration;
}

export interface CreateMatcherTripperOpts extends CreateDefaultTripperOpts {
    name: string;
}

export interface TripperOpts extends CreateMatcherTripperOpts {
    breakerName: string;
}

interface Outcome {
    success: boolean;
    timestamp: number;
}

export enum TripperState {
    closed = "closed",
    half = "half",
    open = "open"
}

/**
 * Maintains state on call outcomes, and encapsulates logic
 * around opening/closing the call circuit
 */
export class Tripper {
    private outcomes: Outcome[] = [];
    private state: TripperState = TripperState.closed;
    private openToHalfTimeout?: NodeJS.Timer;

    constructor(private tripperOpts: TripperOpts) {
        this.startOldOutcomeRemoval();
    }

    /**
     * When we get a failure, if:
     * half-open - immediately go full open
     * closed - open only if this pushes us over threshold
     */
    public recordFailure = () => {
        this.outcomes.push({
            success: false,
            timestamp: Date.now()
        });

        if (
            this.state === TripperState.half ||
            this.currentFailurePercentage() >= this.tripperOpts.threshold
        ) {
            this.open();
        }
    };

    /**
     * When we get a success
     * If half open, we can now go to closed
     */
    public recordSuccess = () => {
        this.outcomes.push({
            success: true,
            timestamp: Date.now()
        });

        if (this.state === TripperState.half) {
            this.state = TripperState.closed;
        }
    };

    public isTripped = () => this.state === TripperState.open;

    public status = () => ({
        currentFailurePercentage: this.currentFailurePercentage(),
        state: this.state
    });

    /**
     * Open the breaker to prevent calls
     * The openToHalfTimeout is kept in local state so that if simultaneous
     * calls cause opening, we don't create multiple timeouts
     */
    private open() {
        this.state = TripperState.open;

        if (this.openToHalfTimeout) {
            clearTimeout(this.openToHalfTimeout);
        }

        this.openToHalfTimeout = setTimeout(() => {
            this.state = TripperState.half;
        }, toMilliseconds(this.tripperOpts.for));
    }

    /**
     * Every outcomeCleanupInterval, look for and remove stored outcomes
     * that are no longer relevant, according to the within measure
     */
    private startOldOutcomeRemoval = () => {
        setTimeout(
            this.startOldOutcomeRemoval.bind(this),
            toMilliseconds(
                this.tripperOpts.outcomeCleanupInterval || seconds(10)
            )
        );
    };

    /**
     * Remove outcomes that are out of the period within which we are interested
     * Returns all outcomes so that other methods can get a snapshot
     */
    private removeOldOutcomes = () => {
        const cutoff = Date.now() - toMilliseconds(this.tripperOpts.within);
        const inScopeOutcomes = this.outcomes.filter(
            ({ timestamp }) => timestamp >= cutoff
        );
        this.outcomes = inScopeOutcomes;
        return inScopeOutcomes;
    };

    /**
     * Calculate the percentage of calls that have failed within the interested
     * period (within)
     */
    private currentFailurePercentage = () => {
        // Get a snapshot of outcomes relevant to now
        const outcomes = this.removeOldOutcomes();

        const [succeeds, fails] = outcomes.reduce(
            ([s, f], { success }) => [
                s + (success ? 1 : 0),
                f + (success ? 0 : 1)
            ],
            [0, 0]
        );

        if (fails === 0) {
            return 0;
        }

        return (100 * fails) / (succeeds + fails);
    };
}
