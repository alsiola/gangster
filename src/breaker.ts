import {
    AsyncFunction,
    TypedEventEmitter,
    TypedEventEmitterInternal
} from "./types";
import { BreakerError, ErrorCode } from "./breaker-error";
import {
    Tripper,
    CreateDefaultTripperOpts,
    CreateMatcherTripperOpts
} from "./tripper";
import { FailTester } from "./fail-tester";
import { BreakerEvents, BreakerEventNames, TripperEventNames } from "./events";
import { EventEmitter } from "events";

interface MatchTripper<T> {
    match: (a: T) => boolean;
}

type MatchTripperCreationOpts<T> = CreateMatcherTripperOpts & MatchTripper<T>;

export interface BreakerInternalOpts<T, U> {
    failTester: FailTester<U>;
    matchTrippers: Array<MatchTripperCreationOpts<T>>;
    defaultTripper: CreateDefaultTripperOpts;
}

export interface BreakerOpts<T, U> extends BreakerInternalOpts<T, U> {
    name: string;
}

export class Breaker<T, U>
    implements TypedEventEmitter<BreakerEvents<T, U>, Breaker<T, U>> {
    private trippers: Array<{ tripper: Tripper } & MatchTripper<T>>;
    private failTester: FailTester<U>;
    private name: string;
    private emitter: TypedEventEmitterInternal<
        BreakerEvents<T, U>,
        Breaker<T, U>
    >;

    constructor(
        private wrappedFunction: AsyncFunction<T, U>,
        { failTester, matchTrippers, defaultTripper, name }: BreakerOpts<T, U>
    ) {
        this.name = name;
        this.trippers = [
            {
                match: () => true,
                tripper: new Tripper({
                    ...defaultTripper,
                    breakerName: name,
                    name: "default"
                })
            },
            ...matchTrippers.map(({ match, ...opts }) => ({
                match,
                tripper: new Tripper({
                    ...opts,
                    breakerName: name
                })
            }))
        ];
        this.failTester = failTester;
        this.emitter = new EventEmitter() as any;
    }

    public on: TypedEventEmitter<BreakerEvents<T, U>, Breaker<T, U>>["on"] = (
        event: any,
        cb: any
    ) => {
        // Tripper events are delegated to the tripper event emitters
        if (event === BreakerEventNames.tripperOpened) {
            this.trippers.forEach(({ tripper }) =>
                tripper.on(TripperEventNames.opened, cb)
            );
            return this;
        }
        this.emitter.on(event, cb);
        return this;
    };

    public status = () => ({
        name: this.name,
        status: this.trippers.map(({ tripper }) => tripper.status())
    });

    /**
     * Attempt to call the wrapped function
     */
    public call = (args: T): Promise<U> => {
        this.emitter.emit(BreakerEventNames.called, { args: args });
        /**
         * Find any trippers that match the arguments, including the default
         * If any are open, this call cannot be allowed to continue
         */
        const openTripper = this.trippers.find(
            ({ tripper, match }) => match(args) && tripper.isTripped()
        );
        if (openTripper) {
            this.emitter.emit(BreakerEventNames.callBlocked, {
                args,
                tripperName: openTripper.tripper.name
            });
            throw new BreakerError(
                "Circuit breaker is open",
                ErrorCode.BreakerOpen
            );
        }

        /**
         * No trippers are open, so we can make the call
         *
         * If it succeeds (i.e. resolves), then we further check that it
         * isn't a failure case with the fail tester - this allows checking
         * e.g. specific HTTP codes, where it may resolve to a 204, but only
         * a 200 would be considered acceptable.
         *
         * If it rejects, then we test the error - in case certain error codes
         * should not be considered failures, e.g. a 404.
         *
         * In both cases, record the success/failure with all trippers that are
         * a match for the arguments
         */
        this.emitter.emit(BreakerEventNames.callAllowed, {
            args
        });
        try {
            return this.wrappedFunction(args).then(result => {
                if (this.failTester.isFailureResult(result)) {
                    this.emitter.emit(BreakerEventNames.invalidResult, {
                        args,
                        result
                    });
                    this.getMatchingTrippers(args).forEach(({ tripper }) =>
                        tripper.recordFailure()
                    );
                    throw new BreakerError(
                        "Invalid function result",
                        ErrorCode.InvalidResult
                    );
                }
                this.emitter.emit(BreakerEventNames.validResult, {
                    args,
                    result
                });
                this.getMatchingTrippers(args).forEach(({ tripper }) =>
                    tripper.recordSuccess()
                );
                return result;
            });
        } catch (error) {
            if (this.failTester.isFailureError(error)) {
                this.emitter.emit(BreakerEventNames.invalidError, {
                    args,
                    error
                });
                this.getMatchingTrippers(args).forEach(({ tripper }) =>
                    tripper.recordFailure()
                );
            } else {
                this.emitter.emit(BreakerEventNames.validError, {
                    args,
                    error
                });
                this.getMatchingTrippers(args).forEach(({ tripper }) =>
                    tripper.recordSuccess()
                );
            }
            throw new BreakerError(
                "Function call failed",
                ErrorCode.FunctionThrew,
                error
            );
        }
    };

    private getMatchingTrippers = (arg: T) => {
        return this.trippers.filter(({ match }) => match(arg));
    };
}
