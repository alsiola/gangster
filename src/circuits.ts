import { Breaker, BreakerOpts, BreakerInternalOpts } from "./breaker";
import { AsyncFunction } from "./types";

export class Circuits {
    private breakers: Breaker<any, any>[] = [];
    constructor(private breakerOpts: BreakerOpts<any, any>) {}

    public createBreaker = <T, U>(f: AsyncFunction<T, U>): Breaker<T, U> => {
        const breaker = new Breaker(f, this.breakerOpts);
        this.breakers.push(breaker);
        return breaker;
    };

    public status = () => {
        return this.breakers.map(breaker => breaker.status());
    };
}
