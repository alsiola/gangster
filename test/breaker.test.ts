import { Breaker } from "../src/breaker";
import { FailTester, seconds } from "../src";
import { delay } from "./helpers";

type TestFunction = (a: TestArgs) => Promise<TestResults>;

enum TestResults {
    good = "good",
    bad = "bad"
}

enum TestArgs {
    good = "good",
    bad = "bad",
    goodVar = "goodVar",
    badVar = "badVar"
}

describe("breaker", () => {
    let testFunction: TestFunction;

    beforeEach(() => {
        testFunction = jest.fn(a => {
            if (a === TestArgs.good || a === TestArgs.goodVar) {
                return Promise.resolve(TestResults.good);
            } else {
                throw TestResults.bad;
            }
        });
    });

    it("allows calls when the default tripper is closed", async () => {
        const breaker = new Breaker(testFunction, {
            name: "test",
            failTester: new FailTester(),
            defaultTripper: {
                for: seconds(1),
                within: seconds(10),
                threshold: 50
            },
            matchTrippers: []
        });

        const result = await breaker.call(TestArgs.good);

        expect(result).toBe(TestResults.good);
    });

    it("prevents calls when the default tripper is open", async () => {
        const breaker = new Breaker(testFunction, {
            name: "test",
            failTester: new FailTester(),
            defaultTripper: {
                for: seconds(1),
                within: seconds(10),
                threshold: -1
            },
            matchTrippers: []
        });

        expect(() => breaker.call(TestArgs.bad)).toThrowError();

        expect(testFunction).toHaveBeenCalledTimes(1);

        expect(() => breaker.call(TestArgs.good)).toThrowError();

        expect(testFunction).toHaveBeenCalledTimes(1);
    });

    it("prevents calls matching arguments when the match tripper is open", async () => {
        const breaker = new Breaker(testFunction, {
            name: "test",
            failTester: new FailTester<TestResults>(),
            defaultTripper: {
                for: seconds(1),
                within: seconds(10),
                threshold: -1
            },
            matchTrippers: [
                {
                    name: "test-match",
                    match: a => a === TestArgs.goodVar || a === TestArgs.badVar,
                    for: seconds(1),
                    within: seconds(10),
                    threshold: -1
                }
            ]
        });

        expect(() => breaker.call(TestArgs.badVar)).toThrowError();

        expect(testFunction).toHaveBeenCalledTimes(1);

        expect(() => breaker.call(TestArgs.goodVar)).toThrowError();

        expect(testFunction).toHaveBeenCalledTimes(1);
    });

    it("allows calls not matching arguments when the match tripper is open", async () => {
        const breaker = new Breaker(testFunction, {
            name: "test",
            failTester: new FailTester<TestResults>(),
            defaultTripper: {
                for: seconds(0.01),
                within: seconds(10),
                threshold: 51
            },
            matchTrippers: [
                {
                    name: "test-match",
                    match: a => a === TestArgs.goodVar || a === TestArgs.badVar,
                    for: seconds(1),
                    within: seconds(10),
                    threshold: -1
                }
            ]
        });

        breaker.call(TestArgs.good);

        expect(() => breaker.call(TestArgs.badVar)).toThrowError();

        await delay(seconds(0.1));

        const result = await breaker.call(TestArgs.good);

        expect(result).toBe(TestResults.good);
    });
});
