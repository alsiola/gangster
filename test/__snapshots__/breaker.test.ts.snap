// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breaker uses failure tester to match errors 1`] = `[Error: Function call failed]`;

exports[`breaker uses failure tester to match results 1`] = `[Error: Invalid function result]`;
